app/goprotoc/goprotoc.go

```diff
@@ -0,0 +1,316 @@
+// Package goprotoc implements the goprotoc command logic.
+package goprotoc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/desc/protoparse"
+)
+
+const protocVersionEmu = "goprotoc 3.5.1"
+
+var (
+	gitSha     = "" // can be replaced by -X linker flag
+	errVersion = errors.New("__version_printed__")
+	errUsage   = errors.New("__usage_printed__")
+)
+
+// Main is the entrypoint for the program.
+func Main() {
+	os.Exit(Run(os.Args, os.Stdin, os.Stdout, os.Stderr))
+}
+
+// Run runs the program and returns the exit code.
+func Run(args []string, stdin io.Reader, stdout io.Writer, stderr io.Writer) int {
+	if err := run(args, stdin, stdout, stderr); err != nil {
+		if message := err.Error(); message != "" {
```

Comment:
This is a bit strange. Have you encountered a case where code generated a blank error?

Reply:
Yes all the time, it's very annoying when it happens, you'd be surprised - people do do `errors.New("")` or `errors.New(someStringTheyExpectToBeNonEmpty)` a decent amount.

Reply:
Strange. I have _never_ seen that. That is unfortunate practice. Though, TBH, it feels like the wrong thing to do to just print nothing but exit with a non-zero code. That would really frustrate me as a user of a command-line tool. You could skip this check, and the result is the same (well, you'd get a single blank line of output instead of nothing -- not materially different to a user).

If you're really concerned about it (FWIW, I wouldn't be), maybe at least do _something_ -- like `fmt.Printf("unexpected error: %#v\n", err)` or panic...

Reply:
I'll update to that - I personally prefer printing nothing, since in theory this is what is intended, but I'll print `"unexpected error"` - in this case, the above will always print `"unexpected error: "`, since we know we have an empty error

Reply:
Done